Registry Access Management (#14300)
docker/docker.github.io: docker-hub/registry-access-management.md

---
description: Registry Access Management
keywords: registry, access, managment
title: Registry Access Management
---

Registry Access Management is a feature available to organizations with a Docker Business subscription. This feature lets organization owners manage the registries that their developers can access while using Docker Desktop.  When using this feature, organization owners can ensure that their developers can only access their trusted registries, such as a secure private registry on Artifactory, thereby reducing the security risks that can occur when developers interact with public registries.

> 

Reviewer: ```suggestion

> **Note**
>
> Registry Access Management is currently offered as a Tech Preview to a closed group of Docker Business customers.
{: .important}

```

### Enforce authentication

To ensure that each org member uses Registry Access Management on their local machine, you can perform the steps below to enforce sign-in under your organization. To do this:

1. Download the latest version of Docker Desktop, and then
2. Create a `registry.json` file.

Download Docker Desktop 4.5 or a later release.

- [Download and install for Windows](/desktop/windows/install/)
- [Download and install for Mac](/desktop/mac/install/)

#### Create a registry json file

Before you create a `registry.json` file, ensure that the developer is a member of at least one organization in Docker Hub. If the registry.json file matches at least one organization the developer is a member of, they can sign into Docker Desktop and access all of their organizations.

**On Windows**

On Windows, you must create a file `C:\ProgramData\DockerDesktop\registry.json` with file permissions that ensure that the developer using Docker Desktop cannot remove or edit the file (that is, only the system administrator can write to the file). The file must be `JSON` and contain one or more organization names in the `allowedOrgs` key.

To create your `registry.json` file on Windows:

> 1. Open Windows Powershell and select Run as Administrator.

Reviewer: ```suggestion
1. Open Windows PowerShell and select **Run as Administrator**.
```

> **Note**
>
> Registry Access Management is currently offered as a Tech Preview to a closed group of Docker Business customers.
{: .important}

## Configure Registry Access Management permissions

To configure Registry Access Management permissions, perform the following steps:

1. Sign into your [Docker Hub](https://hub.docker.com){: target="_blank" rel="noopener" class="_"} account as an organization owner.
2. Select an organization, navigate to the **Settings** tab on the **Organizations** page and click **Org Permissions**.
3. Enable Registry Access Management to set the permissions for your registry.

> **Note**
>
> When enabled, the Docker Hub registry is set by default, however you can also restrict this registry for your developers.

> 4. Click **Add** and enter your registry details in the applicable fields, and click **Create** to add the registry to your list.

Reviewer 2: On the preview page I am seeing items 4 and 5 incorrectly numbered as items 1 and 2. https://deploy-preview-14300--docsdocker.netlify.app/docker-hub/registry-access-management/

Reviewer: @jerae-duffin Could you fix this. The note after 1 and 2 needs indenting

Author: Updated

To create your `registry.json` file on macOS:

1. Navigate to Visual Studio Code or any text editor of your choice.
2. Enter one or more organization names in the `allowedOrgs` key and save it in your Documents.

    For example:

    ```json
    {
     "allowedOrgs": ["myorg"]
    }
    ```

 3. Open a new terminal and type the following command:

    `sudo mkdir -p /Library/Application\ Support/com.docker.docker`

    Note: if prompted, type your password associated with your local computer.

4. Type the following command:

    `sudo cp Documents/registry.json /Library/Application\ Support/com.docker.docker/registry.json`

### Verify the restrictions

   After you’ve created the registry.json file and deployed it onto the developers’ machines, you can verify whether the changes have taken effect by asking the developers to start Docker Desktop.

>    If the configuration is successful, Docker Desktop prompts the developer to authenticate using the organization credentials on start. If the developer fails to authenticate, our authenticaes as a developer in the wrong organization they will see an error message, and they will be denied access to Docker Desktop.

Reviewer 3: "or authenticates" not "our authenticaes"

Reviewer: Good catch!

Author: Updated, thank you

---
description: Registry Access Management
keywords: registry, access, managment
title: Registry Access Management
---

> Registry Access Management is a new feature available to organizations with a Docker Business subscription. This feature lets organization owners manage the registries that their users can access while using Docker Desktop.  When using this feature, organization owners can ensure that their users can only access their trusted registries, such as a secure private registry on Artifactory, thereby reducing the security risks that can occur when users interact with public registries.

Reviewer 3: Do we want to say that it's "new" within the text? Will we remember to remove that in a few months time? I would just delete "a new feature" and say "is available to". But you're the expert so I'm happy to be overruled.

Author: Thanks Stephen! Good point, I removed  "new"

## Configure Registry Access Management permissions

To configure Registry Access Management permissions, perform the following steps:

> 1. Sign into your [Docker Hub](https://hub.docker.com) account as an organization owner.

Reviewer: Could you add the `target blank` attribute to open Docker Hub in a new tab?

2. Select an organization, navigate to the **Settings** tab on the **Organizations** page and click **Org Permissions**.
3. Enable Registry Access Management to set the permissions for your registry.

> **Note**
>
> When enabled, the Docker Hub registry is set by default, however you can also restrict this registry for your developers.

4. Click **Add** and enter your registry details in the applicable fields, and click **Create** to add the registry to your list.
5. Verify that the registry appears in your list and click **Save & Apply**.  You can verify that your changes are saved in the Activity tab.

  > **Note**
  >

>   > Once you add a registry, it can take a maximum of 24 hours for the changes to be enforced on your developers’ machines. If you want to apply the changes sooner, you must force a Docker logout on your developers’ machine and have the developer re-authenticate for Docker Desktop.  Also, there is no limit on the number of registries you can add. See the Caveats sections to learn more about limitations  when using this feature.

Reviewer: ```suggestion
  > Once you add a registry, it can take up to 24 hours for the changes to be enforced on your developers’ machines. If you want to apply the changes sooner, you must force a Docker logout on your developers’ machine and have the developers re-authenticate for Docker Desktop.  Also, there is no limit on the number of registries you can add. See the Caveats sections to learn more about limitations  when using this feature.
```

Reviewer: Could you add a link to the Caveats section?

Author: Updated

### Enforce authentication

> To ensure that each org member uses Registry Access Management on their local machine, you can perform the following steps below to enforce sign-in under your organization. To do this:

Reviewer: change to "following steps" or "steps below"

Author: Updated

1. Download the latest version of Docker Desktop, and then
2. Create a `registry.json` file.

Download Docker Desktop 4.5 or a later release.

- [Download and install for Windows](/desktop/windows/install/)
- [Download and install for Mac](/desktop/mac/install/)

#### Create a registry json file

> Before you create a `registry.json` file, ensure that the developer is a member of at least one organization in Docker Hub. If the registry.json file matches at least one organization the developer is a member of, they can sign into Docker Desktop and access all of their organizations.

Reviewer: format: `registry.json`

Author: Updated

**On Windows**

On Windows, you must create a file `C:\ProgramData\DockerDesktop\registry.json` with file permissions that ensure that the developer using Docker Desktop cannot remove or edit the file (that is, only the system administrator can write to the file). The file must be `JSON` and contain one or more organization names in the `allowedOrgs` key.

> To create your `registry.json` file on Windows:

Reviewer 3: I'm not a PowerShell expert, so I might be wrong, but I'm confused by these commands. Step 3 doesn't seem to make sense and what does Step 4 do? Have you tested these instructions?

Reviewer 3: Oh, I think I see what's happened now. The instructions at https://docs.docker.com/docker-hub/configure-sign-in/#create-a-registryjson-file make more sense. It looks like this copy has either been mis-edited from there, or the ones there have received later corrections.

Reviewer 4: It must be "exactly one organization name" in the `allowedOrgs` key. We had to remove the ability to list multiple organizations there. Unfortunately the value is still syntactically a list (but now a list with exactly one element)

Author: Hey @stephen-turner Updated. Aligned with the testing doc, as well as the create-a-registryjson-file link in Docs

> 1. Open Windows Powershell and select Run as Administrator.

Reviewer: Format **Run as Administrator*.

2. Type the following command: `cd /ProgramData/DockerDesktop/`

> 3. In Notepad, type `registry.json` and enter the Docker Hub organization that the user belongs to in `allowedOrgs` key and click Save.

Reviewer: Format **Save**

Author: Updated

    For example:

    ```json
    {
    "allowedOrgs": ["myorg"]
    }
    ```

4. Navigate to Powershell and type ```start .```

> Congratulations! You have just created the registry.json file.

Reviewer 3: It's not that big a deal! I suggest just removing this line.

Author: Updated

**On macOS**:

> On macOS, you must create a file at `/Library/Application Support/com.docker.docker/registry.json` with file permissions that ensure that the developer using Docker Desktop cannot remove or edit the file (that is, only the system administrator can write to the file). The file must be of type JSON and contain the name of the Docker Hub organization names in the allowedOrgs key.

Reviewer 4: "and contain the name of the Docker Hub organization in the allowedOrgs key" (one organization name, rather than multiple)

Author: Thanks @djs55 Updated

To create your `registry.json` file on macOS:

> 1. Navigate to VS Code or any text editor of your choice.

Reviewer 3: Visual Studio Code

Author: Thanks @stephen-turner Updated

2. Enter one or more organization names in the `allowedOrgs` key and save it in your Documents.

    For example:

    ```json
    {
     "allowedOrgs": ["myorg"]
    }
    ```

 3. Open a new terminal and type the following command:

    `sudo mkdir -p /Library/Application\ Support/com.docker.docker`

    Note: if prompted, type your password associated with your local computer.

4. Type the following command:

    `sudo cp Documents/registry.json /Library/Application\ Support/com.docker.docker/registry.json`

> Congratulations! You have just created the `registry.json` file.

Reviewer 3: Again, delete this line.

Author: Updated

### Verify the restrictions

>    After you’ve created the registry.json file and deployed it onto the users’ machines, you can verify whether the changes have taken effect by asking users to start Docker Desktop.

Reviewer: I see that we use both developers and users here. Suggest we use either developers or users to be consistent.

>    If the configuration is successful, Docker Desktop prompts the user to authenticate using the organization credentials on start. If the user fails to authenticate, they will see an error message, and they will be denied access to Docker Desktop.

Reviewer 4: should we also add "If the user fails to authenticate, or authenticates as a user in the wrong organization, they will see..." (this confused me for a while because my password manager filled the password in and yet I was still denied, I didn't spot the subtle difference in the UI message)

Author: Updated

### Caveats

  There are certain limitations when using Registry Access Management; they are as follows:

  * Windows image pulls, and image builds are not restricted
  * Builds such as `docker buildx` using a Kubernetes driver are not restricted
  * Builds such as `docker buildx` using a custom docker-container driver are not restricted
  * Blocking is DNS-based; you must use a registry's access control mechanisms to distinguish between “push” and “pull”
  * You must disable HTTP proxy or use a corporate proxy which also blocks the registries
  * WSL 2 requires at least a 5.4 series Linux kernel  (this does not apply to earlier Linux kernel series)
 * Under the WSL 2 network, traffic from all Linux distributions is restricted (this will be resolved in the updated 5.15 series Linux kernel)

>   Also, Registry Access Management operates on the level of hosts, not IP addresses, that can be bypassed if a user manipulates their domain resolution, such as by running Docker against a local proxy or modifying their operating system's `sts` file. Blocking these forms of manipulation is outside the remit of Docker Desktop.

Reviewer 3: I can't parse the first sentence.

To create your `registry.json` file on macOS:

1. Navigate to Visual Studio Code or any text editor of your choice.
2. Enter one or more organization names in the `allowedOrgs` key and save it in your Documents.

    For example:

    ```json
    {
     "allowedOrgs": ["myorg"]
    }
    ```

 3. Open a new terminal and type the following command:

    `sudo mkdir -p /Library/Application\ Support/com.docker.docker`

    Note: if prompted, type your password associated with your local computer.

4. Type the following command:

    `sudo cp Documents/registry.json /Library/Application\ Support/com.docker.docker/registry.json`

### Verify the restrictions

   After you’ve created the registry.json file and deployed it onto the developers’ machines, you can verify whether the changes have taken effect by asking the developers to start Docker Desktop.

>    If the configuration is successful, Docker Desktop prompts the developer to authenticate using the organization credentials on start. If the developer fails to authenticate, or authenticaes as a developer in the wrong organization they will see an error message, and they will be denied access to Docker Desktop.

Reviewer: typo: authenticates

Author: Thanks Usha! Updated
